### PR TITLE
Big endian support

### PR DESCRIPTION
**Pull request by AJenbo**
This should make the game fully functional on big-endian CPUs, paving the way for Amiga, PS3, Xbox 360 support, and let's not forget the original PowerPC Mac that got official support in 1998.

### CODE REVIEW

File: Source/loadsave.cpp [hunk 1]
```diff
@@ -215,32 +215,53 @@ void CopyBytes(const void *src, const int n, void *dst)
...
+	memcpy(&buf, src, 4);
+	tbuff += 4;
+	buf = SwapLE32(buf);
+	memcpy(dst, &buf, 4);
 }
 
 void CopyInts(const void *src, const int n, void *dst)
 {
-	CopyBytes(src, 4 * n, dst);
+	unsigned int *s=src, *d=dst;
```
Comment (AJenbo) [inline Source/loadsave.cpp]: This reassignment from `const` breaks Windows and Mac builds.

**Comment by AJenbo** [comment]
@Samuel-DEVULDER could you check the changes I did to load/save in order to fix accessing misaligned addresses?

**Comment by AJenbo** [comment]
@glebm do you think it's worth adding the changes to misc_io.cpp?

File: SourceX/miniwin/misc_io.cpp [hunk 2]
```diff
@@ -33,11 +33,13 @@ HANDLE CreateFileA(LPCSTR lpFileName, DWORD dwDesiredAccess, DWORD dwShareMode,
 	file->path = name;
 	if (dwCreationDisposition == DVL_OPEN_EXISTING) {
 		// read contents of existing file into buffer
-		std::ifstream filestream(file->path, std::ios::binary);
-		if (!filestream.fail()) {
-			file->buf.insert(file->buf.begin(),
-			    std::istreambuf_iterator<char>(filestream),
-			    std::istreambuf_iterator<char>());
+		std::ifstream filestream(file->path, std::ios::in | std::ios::binary);
```
Comment (glebm) [inline SourceX/miniwin/misc_io.cpp]: Open with `std::ios::ate` to remove the need for the first `seekg` call

File: SourceX/miniwin/misc_io.cpp [hunk 3]
```diff
@@ -33,11 +33,13 @@ HANDLE CreateFileA(LPCSTR lpFileName, DWORD dwDesiredAccess, DWORD dwShareMode,
...
-			file->buf.insert(file->buf.begin(),
-			    std::istreambuf_iterator<char>(filestream),
-			    std::istreambuf_iterator<char>());
+		std::ifstream filestream(file->path, std::ios::in | std::ios::binary);
+		if (filestream) {
+			filestream.seekg(0, std::ios::end);
+			file->buf.resize(filestream.tellg());
+			filestream.seekg(0, std::ios::beg);
+			filestream.read(&file->buf[0], file->buf.size());
+			filestream.close();
```
Comment (glebm) [inline SourceX/miniwin/misc_io.cpp]: No need to `close`, the file stream is destroyed right after.

File: SourceX/miniwin/misc_io.cpp (hunk 2 above)
Comment (AJenbo) [reply SourceX/miniwin/misc_io.cpp]: Both `seekg` calls?

**Comment by glebm** [comment]
> @glebm do you think it's worth adding the changes to misc_io.cpp?

Totally, a tiny bit faster, probably won't matter much but it's a better idiom.

File: SourceX/miniwin/misc_io.cpp [hunk 4]
```diff
@@ -33,11 +33,13 @@ HANDLE CreateFileA(LPCSTR lpFileName, DWORD dwDesiredAccess, DWORD dwShareMode,
...
-		std::ifstream filestream(file->path, std::ios::binary);
-		if (!filestream.fail()) {
-			file->buf.insert(file->buf.begin(),
-			    std::istreambuf_iterator<char>(filestream),
-			    std::istreambuf_iterator<char>());
+		std::ifstream filestream(file->path, std::ios::in | std::ios::binary);
+		if (filestream) {
+			filestream.seekg(0, std::ios::end);
+			file->buf.resize(filestream.tellg());
+			filestream.seekg(0, std::ios::beg);
```
Comment (glebm) [inline SourceX/miniwin/misc_io.cpp]: Not 100% sure, but I think can just be `filestream.seekg(0)`
http://www.cplusplus.com/reference/istream/istream/seekg/

**Comment by AJenbo** [comment]
> Last: is there any BigEndian platform requiring alignment of shorts? Ever since the 68020 Motorola supports any alignment.

My PowerPC 7400 segfaults because of the alignment issues.

> I wonder: why in the first place would the src or dst be mis-aligned? These are supposed to contain short, so be correctly aligned in the first place even in bytes inside are in wrong order.

One of them is just a stream of bytes from the save file so I'm thinking this probably isn't aligned. The current code could do with some refactoring to have it be more type aware. The issue I was having with loading booleans would be nice to solve, and would give us back some extra bytes in the save files. But yeah let's wait with that till later.

File: SourceX/miniwin/misc_io.cpp (hunk 2 above)
Comment (glebm) [reply SourceX/miniwin/misc_io.cpp]: Only the first `seekg`. `std::ios::ate` takes you to the end of the file on opening.

**Pull request merged**